### PR TITLE
Replace GetService with GetServiceAsync calls in Visual Studio

### LINKED ISSUE #3784: Replace GetService with GetServiceAsync everywhere in NuGet.Clients
Follow up on #3775.

Recommended practice of acquiring VS services is always calling `GetServiceAsync` to avoid the high risk of deadlock. All the occurrences of `GetService` and `ServiceLocator` should be replaced.

### PR DESCRIPTION
**Pull request by erdembayar**
Fixes: https://github.com/NuGet/Home/issues/3784

Regression? No
## Description
Here addressing "Recommended practice of acquiring VS services is always calling GetServiceAsync to avoid the high risk of deadlock". It's incremental effort and this is part 2. Here reduced sync calls for all remaining GetService calls not addressed in [PR#3766](https://github.com/NuGet/NuGet.Client/pull/3766).

Replaced non-async with async calls in following files:
src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourceOptionsPage.cs
src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs
src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs
src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs
src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs

Removed dead(no longer relevant) codes from following files due to above refactoring:
src/NuGet.Clients/NuGet.VisualStudio.Common/GlobalSuppressions.cs
src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceProviderExtensions.cs
src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by zivkan** [comment]
This PR has the string `.Run(` 11 times.  Are you sure this achieves anything in relation to reducing risk of deadlock? JTF will only allow async tasks "related" to the delegate/action passed to `Run`. Any other async task, or sync code, that is trying to switch to the UI thread is still going to be blocked until `.Run(...)` returns.

Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.

Secondly, in a few places you removed `IServiceProvider` or `IAsyncServiceProvider` being passed in method calls, and instead use a global/static async service provider. This makes the code less testable, as it's no longer possible to easily pass in a mock service provider that resolves mock services. Even if the static service provider allows tests to override it, it means that tests can't be run in parallel. Even though the classes you modified aren't currently being tested, so it "doesn't matter", it raises the barrier for anyone who wants to start testing the class in the future, so I don't think it's a good change.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: I think there's a better way to do this. 

The `GetService` call that used to be here is a consequence of this getting invoked on a synchronous codepath.

Changing to JTF.Run remove the GetService call, but doesn't really do anything in the grandscheme of improving perf. 

To improve perf, you should go another layer up. 

For example, this is *only* in https://github.com/NuGet/NuGet.Client/blob/a1fa3375d91d95e617510480f4c3b28324563f94/src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs#L521. 

Which is already an async codepath, so the *correct* thing to do here would be to call GetServiceAsync *directly*. 

Same applies to *all* situations.

**Comment by erdembayar** [comment]
> Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.
> 

It's just body of initial issue. https://github.com/NuGet/Home/issues/3784#issue-185520889
Actually I don't have any good guide, I'm just focusing on replace all '.GetReplace' vs '.GetReplaceAsync' which are listed by @dominoFire [here](https://github.com/NuGet/Home/issues/3784#issue-185520889).

**Review by nkolev92 — COMMENTED** [review]
I think we haven't clarified the expectations for this task too well. 

I left a comment that hopefully improves that. 
I'd be happy to have an offline chat tomorrow if need be.

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Ok. I addressed PR comment concerns and I tried to make it async all the way up.
Also I introduced some async event delegate handlers based on existing pattern we already have:
https://github.com/NuGet/NuGet.Client/blob/f754a0da911a07cfac987df0fbdf5518692e3d5a/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageMetadataControl.xaml.cs#L41-L49

Almost all 3 cases they're working on UI so I don't believe they introduce any timing bug since they're already fire and forget type things.

1. PowerConsoleToolWindow.cs : Update menu and commands of VS. 
2. PackageSourcesOptionsControl.cs : Open browse explorer in "Package Source Option"
image

3. OptionsPageActivator.cs : Open "General" and "Package sources"
image

image

**Comment by erdembayar** [comment]
> This PR has the string `.Run(` 11 times. Are you sure this achieves anything in relation to reducing risk of deadlock? JTF will only allow async tasks "related" to the delegate/action passed to `Run`. Any other async task, or sync code, that is trying to switch to the UI thread is still going to be blocked until `.Run(...)` returns.
> 
Thank you for your review.

Taken your concern in consideration. Unless at least one direct async call benefit I'm not adding async wrapped in `.Run()` by default.

> Also, can you post a link to the source of the recommended practice? Since you put it in double quotes, I assumed you copy/pasted it from somewhere, but searching for it online only find your own PR. I'm curious if the recommendation gives more details.
> 
Sorry, I don't have recommended practice. That was just body text of this [issue](https://github.com/NuGet/Home/issues/3784#issue-185520889).

> Secondly, in a few places you removed `IServiceProvider` or `IAsyncServiceProvider` being passed in method calls, and instead use a global/static async service provider. This makes the code less testable, as it's no longer possible to easily pass in a mock service provider that resolves mock services. Even if the static service provider allows tests to override it, it means that tests can't be run in parallel. Even though the classes you modified aren't currently being tested, so it "doesn't matter", it raises the barrier for anyone who wants to start testing the class in the future, so I don't think it's a good change.

Taken your advise and replacing constructor based service provider with global service provider. But I did replace non-async service provider with async service provider through constructor injection.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: Given that this is called in 1 place only, you don't need the extra allocation. Just make the async provider call in the calling method instead.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: You don't really need the extra delegate: 

```cs
                        NuGetUIThreadHelper.JoinableTaskFactory.Run(OnDispatcherStartCompletedAsync);
``` 

The above should do the job as well.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: You don't need an extra delegate here. 
See above comment.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: Given that this is now async, this check is incorrect. 

See rule #1 https://github.com/microsoft/vs-threading/blob/main/doc/threading_rules.md

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs:30]: Given the scenario, It might not matter too much, but why did you remove the caching of the DTE object here?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs]: VS threading rule #1.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs:33]: Was the removal of the DTE caching intentional?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs]: vs threading rule #1.

**Review by nkolev92 — COMMENTED** [review]
Looks a lot better. 

A couple of VS rule violations to be addressed.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: Ok. Addressed.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: Ok. Addressed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: Ok. Addressed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Options/PackageSourcesOptionsControl.cs]: Ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VsCommonOperations.cs:30]: Sorry, not intentional. Replaced with AsyncLazy, please check if implementation is correct.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/OptionsPageActivator.cs]: Ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/SourceControl/VSSourceControlManagerProvider.cs:33]: Sorry, not intentional. Replaced with AsyncLazy, please check if implementation is correct.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs
Comment (erdembayar) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/VsTemplateWizard.cs]: ok. Replaced with `await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();`

**Review by erdembayar — COMMENTED** [review]
@nkolev92 
Thank you for your review. I addressed your comment please check.

File: src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.Console/PowerConsoleToolWindow.cs]: You have an extra delegate here.

**Review by nkolev92 — APPROVED** [review]

**Comment by erdembayar** [comment]
@NuGet/nuget-client 
Please review. Need 1 more review  approval.

**Pull request merged**
